File: go/csv_grammar_test.go
package jsonic

import (
	"reflect"
	"testing"
)

// Minimal CSV grammar built directly on the jsonic parser API.
// Mirrors test/csv-grammar.test.js — the grammar treats comma-separated
// values as cells, newline-separated rows as records, and single tokens
// (text, number, string, keyword) as cell values. Empty cells become "";
// empty rows are dropped.
func makeCSV() *Jsonic {
	j := Make()

	// pushBack keeps the replaced-rule chain in sync so the parent rule
	// always sees the latest slice through r.Parent.Child.Node. Go's
	// append may reallocate, which means a bare tail-call replacement
	// would leave the original row.Node stale — unlike JS arrays.
	pushBack := func(r *Rule) {
		if r.Parent != NoRule && r.Parent != nil &&
			r.Parent.Child != NoRule && r.Parent.Child != nil {
			r.Parent.Child.Node = r.Node
		}
	}

	appendCell := func(r *Rule, val any) {
		arr, _ := r.Node.([]any)
		r.Node = append(arr, val)
		pushBack(r)
	}

	// csv: outer list of rows. Fresh bo resets Node for each new parse.
	j.Rule("csv", func(rs *RuleSpec) {
		rs.BO = []StateAction{func(r *Rule, ctx *Context) {
			r.Node = []any{}
		}}
		rs.Open = []*AltSpec{
			{S: [][]Tin{{TinZZ}}},
			{P: "row"},
		}
		rs.Close = []*AltSpec{
			{S: [][]Tin{{TinLN}, {TinZZ}}},
			{S: [][]Tin{{TinLN}}, R: "csvcont"},
			{S: [][]Tin{{TinZZ}}},
		}
		rs.BC = []StateAction{func(r *Rule, ctx *Context) {
			if cells, ok := r.Child.Node.([]any); ok && len(cells) > 0 {
				outer, _ := r.Node.([]any)
				r.Node = append(outer, cells)
			}
		}}
	})

	// csvcont: tail-call sibling of csv. Inherits the outer-list node so
	// the replace chain carries the rows accumulated so far.
	j.Rule("csvcont", func(rs *RuleSpec) {
		rs.Open = []*AltSpec{
			{S: [][]Tin{{TinZZ}}},
			{P: "row"},
		}
		rs.Close = []*AltSpec{
			{S: [][]Tin{{TinLN}, {TinZZ}}},
			{S: [][]Tin{{TinLN}}, R: "csvcont"},
			{S: [][]Tin{{TinZZ}}},
		}
		rs.BC = []StateAction{func(r *Rule, ctx *Context) {
			if cells, ok := r.Child.Node.([]any); ok && len(cells) > 0 {
				outer, _ := r.Node.([]any)
				r.Node = append(outer, cells)
			}
		}}
	})

	// row: handles the first cell (initialising the row slice) then hands
	// the continuation to rowcont. Row-ending tokens at open produce an
	// empty row, which csv.bc drops.
	j.Rule("row", func(rs *RuleSpec) {
		rs.Open = []*AltSpec{
			{S: [][]Tin{TinSetVAL}, A: func(r *Rule, ctx *Context) {
				r.Node = []any{r.O[0].Val}
			}},
			{S: [][]Tin{{TinCA}}, B: 1, A: func(r *Rule, ctx *Context) {
				r.Node = []any{""}
			}},
			{S: [][]Tin{{TinLN}}, B: 1, A: func(r *Rule, ctx *Context) {
				r.Node = []any{}
			}},
			{S: [][]Tin{{TinZZ}}, B: 1, A: func(r *Rule, ctx *Context) {
				r.Node = []any{}
			}},
		}
		rs.Close = []*AltSpec{
			{S: [][]Tin{{TinCA}}, R: "rowcont"},
			{S: [][]Tin{{TinLN}}, B: 1},
			{S: [][]Tin{{TinZZ}}, B: 1},
		}
	})

	// rowcont: continues appending cells into the row slice. pushBack
	// keeps row.Node (the node parent rule reads) synced after append.
	j.Rule("rowcont", func(rs *RuleSpec) {
		rs.Open = []*AltSpec{
			{S: [][]Tin{TinSetVAL}, A: func(r *Rule, ctx *Context) {
				appendCell(r, r.O[0].Val)
			}},
			{S: [][]Tin{{TinCA}}, B: 1, A: func(r *Rule, ctx *Context) {
				appendCell(r, "")
			}},
			{S: [][]Tin{{TinLN}}, B: 1, A: func(r *Rule, ctx *Context) {
				appendCell(r, "")
			}},
			{S: [][]Tin{{TinZZ}}, B: 1, A: func(r *Rule, ctx *Context) {
				appendCell(r, "")
			}},
		}
		rs.Close = []*AltSpec{
			{S: [][]Tin{{TinCA}}, R: "rowcont"},
			{S: [][]Tin{{TinLN}}, B: 1},
			{S: [][]Tin{{TinZZ}}, B: 1},
		}
	})

	// Select the custom start rule and drop jsonic-only extensions.
	// Keep #SP and #CM in IGNORE but let #LN reach the parser.
	j.SetOptions(Options{
		Rule: &RuleOptions{Start: "csv", Exclude: "jsonic,imp"},
		Lex:  &LexOptions{EmptyResult: []any{}},
		TokenSet: map[string][]string{
			"IGNORE": {"#SP", "#CM"},
		},
	})

	return j
}

func runCSV(t *testing.T, name, src string, want []any) {
	t.Helper()
	j := makeCSV()
	got, err := j.Parse(src)
	if err != nil {
		t.Fatalf("%s: Parse(%q) error: %v", name, src, err)
	}
	if !reflect.DeepEqual(got, want) {
		t.Errorf("%s: Parse(%q)\n  got:      %#v\n  expected: %#v",
			name, src, got, want)
	}
}

func TestCSVEmptyInput(t *testing.T) {
	runCSV(t, "empty-input", "", []any{})
}

func TestCSVSingleRow(t *testing.T) {
	runCSV(t, "single-row", "a,b,c",
		[]any{[]any{"a", "b", "c"}})
}

func TestCSVMultipleRows(t *testing.T) {
	runCSV(t, "multiple-rows", "a,b\nc,d",
		[]any{[]any{"a", "b"}, []any{"c", "d"}})
}

func TestCSVTrailingNewline(t *testing.T) {
	runCSV(t, "trailing-newline", "a,b,c\n",
		[]any{[]any{"a", "b", "c"}})
}

func TestCSVBlankLinesSkipped(t *testing.T) {
	runCSV(t, "blank-lines", "a,b\n\nc,d\n",
		[]any{[]any{"a", "b"}, []any{"c", "d"}})
}

func TestCSVNumbersParsed(t *testing.T) {
	runCSV(t, "numbers", "1,2,3",
		[]any{[]any{float64(1), float64(2), float64(3)}})
}

func TestCSVQuotedStrings(t *testing.T) {
	runCSV(t, "quoted", `"hello","world"`,
		[]any{[]any{"hello", "world"}})
}

func TestCSVMixedTypes(t *testing.T) {
	runCSV(t, "mixed", `a,1,"x",true`,
		[]any{[]any{"a", float64(1), "x", true}})
}

func TestCSVEmptyLeadingField(t *testing.T) {
	runCSV(t, "leading-empty", ",a,b",
		[]any{[]any{"", "a", "b"}})
}

func TestCSVEmptyMiddleField(t *testing.T) {
	runCSV(t, "middle-empty", "a,,b",
		[]any{[]any{"a", "", "b"}})
}

func TestCSVEmptyTrailingField(t *testing.T) {
	runCSV(t, "trailing-empty", "a,b,",
		[]any{[]any{"a", "b", ""}})
}

func TestCSVSingleCellRow(t *testing.T) {
	runCSV(t, "single-cell", "x\ny",
		[]any{[]any{"x"}, []any{"y"}})
}

func TestCSVKeywords(t *testing.T) {
	runCSV(t, "keywords", "true,false,null",
		[]any{[]any{true, false, nil}})
}

File: test/csv-grammar.test.js
/* Copyright (c) 2013-2026 Richard Rodger and other contributors, MIT License */
'use strict'

// Minimal CSV grammar expressed directly against the jsonic parser API.
// The grammar is intentionally simple: comma-separated values, newline-separated
// rows, single-token fields (numbers, strings, keywords, and unquoted text).
// Empty fields are emitted as empty strings; empty rows are skipped.

const { describe, it } = require('node:test')
const assert = require('node:assert')

const { Jsonic } = require('..')

function makeCsv() {
  const j = Jsonic.make({
    tokenSet: {
      // Default IGNORE is [#SP, #LN, #CM]. Drop #LN so newlines become
      // parsing tokens; keep space and comment in IGNORE.
      IGNORE: [undefined, null, undefined],
    },
    rule: {
      start: 'csv',
      // Disable jsonic-only extensions (implicit maps, path-diving, etc.).
      exclude: 'jsonic,imp',
    },
    lex: {
      // Empty source should return [] (an empty CSV) rather than undefined.
      emptyResult: [],
    },
  })

  const { CA, LN, ZZ } = j.token
  const { VAL } = j.tokenSet

  // Top-level: collect rows into an array.
  j.rule('csv', (rs) => {
    rs.bo((r) => { r.node = [] })
      .open([
        { s: [ZZ] },
        { p: 'row' },
      ])
      .close([
        { s: [LN, ZZ] },
        { s: [LN], r: 'csvcont' },
        { s: [ZZ] },
      ])
      .bc((r) => {
        if (Array.isArray(r.child.node) && r.child.node.length > 0) {
          r.node.push(r.child.node)
        }
      })
  })

  // csv continuation: same as csv but without the node-resetting bo so the
  // shared outer list survives the tail-call replace.
  j.rule('csvcont', (rs) => {
    rs.open([
      { s: [ZZ] },
      { p: 'row' },
    ])
    .close([
      { s: [LN, ZZ] },
      { s: [LN], r: 'csvcont' },
      { s: [ZZ] },
    ])
    .bc((r) => {
      if (Array.isArray(r.child.node) && r.child.node.length > 0) {
        r.node.push(r.child.node)
      }
    })
  })

  // Row: first cell initialises r.node with its value so that subsequent
  // rowcont iterations can push into the same array. An immediate row-ending
  // token (LN/ZZ) produces an empty row (dropped by csv.bc).
  j.rule('row', (rs) => {
    rs.open([
      { s: [VAL], a: (r) => { r.node = [r.o0.val] } },
      { s: [CA], b: 1, a: (r) => { r.node = [''] } },
      { s: [LN], b: 1, a: (r) => { r.node = [] } },
      { s: [ZZ], b: 1, a: (r) => { r.node = [] } },
    ])
    .close([
      { s: [CA], r: 'rowcont' },
      { s: [LN], b: 1 },
      { s: [ZZ], b: 1 },
    ])
  })

  // Row continuation: push each subsequent cell onto the shared row array.
  j.rule('rowcont', (rs) => {
    rs.open([
      { s: [VAL], a: (r) => { r.node.push(r.o0.val) } },
      { s: [CA], b: 1, a: (r) => { r.node.push('') } },
      { s: [LN], b: 1, a: (r) => { r.node.push('') } },
      { s: [ZZ], b: 1, a: (r) => { r.node.push('') } },
    ])
    .close([
      { s: [CA], r: 'rowcont' },
      { s: [LN], b: 1 },
      { s: [ZZ], b: 1 },
    ])
  })

  return j
}

describe('csv-grammar', () => {
  const csv = makeCsv()

  it('empty-input', () => {
    assert.deepEqual(csv(''), [])
  })

  it('single-row', () => {
    assert.deepEqual(csv('a,b,c'), [['a', 'b', 'c']])
  })

  it('multiple-rows', () => {
    assert.deepEqual(csv('a,b\nc,d'), [['a', 'b'], ['c', 'd']])
  })

  it('trailing-newline', () => {
    assert.deepEqual(csv('a,b,c\n'), [['a', 'b', 'c']])
  })

  it('blank-lines-skipped', () => {
    assert.deepEqual(csv('a,b\n\nc,d\n'), [['a', 'b'], ['c', 'd']])
  })

  it('numbers-are-parsed', () => {
    assert.deepEqual(csv('1,2,3'), [[1, 2, 3]])
  })

  it('quoted-strings', () => {
    assert.deepEqual(csv('"hello","world"'), [['hello', 'world']])
  })

  it('mixed-types', () => {
    assert.deepEqual(csv('a,1,"x",true'), [['a', 1, 'x', true]])
  })

  it('empty-leading-field', () => {
    assert.deepEqual(csv(',a,b'), [['', 'a', 'b']])
  })

  it('empty-middle-field', () => {
    assert.deepEqual(csv('a,,b'), [['a', '', 'b']])
  })

  it('empty-trailing-field', () => {
    assert.deepEqual(csv('a,b,'), [['a', 'b', '']])
  })

  it('single-cell-row', () => {
    assert.deepEqual(csv('x\ny'), [['x'], ['y']])
  })

  it('keywords-recognised', () => {
    assert.deepEqual(csv('true,false,null'), [[true, false, null]])
  })
})
